netty/src/main/java/io/grpc/transport/netty/WriteQueue.java

```diff
@@ -0,0 +1,162 @@
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.transport.netty;
+
+import com.google.common.base.Preconditions;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelPromise;
+
+import java.util.ArrayDeque;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * A queue of pending writes to a {@link Channel} that is flushed as a single unit.
+ */
+class WriteQueue {
+
+  private static final int DEQUE_CHUNK_SIZE = 128;
+
+  /**
+   * {@link Runnable} used to schedule work onto the tail of the event loop.
+   */
+  private final Runnable later = new Runnable() {
+    @Override
+    public void run() {
+      flush();
+    }
+  };
+
+  private final Channel channel;
+  private final BlockingQueue<Runnable> queue;
+  private final AtomicBoolean scheduled = new AtomicBoolean();
+
+  /**
+   * ArrayDeque to copy queue into when flushing in event loop.
+   */
+  private final ArrayDeque<Runnable> writeChunk =
+      new ArrayDeque<Runnable>(DEQUE_CHUNK_SIZE);
+
+  public WriteQueue(Channel channel) {
+    this.channel = Preconditions.checkNotNull(channel, "channel");
+    queue = new LinkedBlockingQueue<Runnable>();
+  }
+
+  /**
+   * Schedule a flush on the channel.
+   */
+  void scheduleFlush() {
+    if (scheduled.compareAndSet(false, true)) {
+      // Add the queue to the tail of the event loop so writes will be executed immediately
+      // inside the event loop. Note DO NOT do channel.write outside the event loop as
+      // it will not wake up immediately without a flush.
+      channel.eventLoop().execute(later);
+    }
+  }
+
+  /**
+   * Enqueue a write command on the channel.
+   *
+   * @param command a write to be executed on the channel.
+   * @param flush true if a flush of the write should be schedule, false if a later call to
+   *              enqueue will schedule the flush.
+   */
+  void enqueue(Object command, boolean flush) {
+    enqueue(command, channel.newPromise(), flush);
```

Comment:
"Easy enough" is not the bar to throw things into an existing PR. I'm more concerned about the additional risk of the change and having a PR be cohesive. I'd like it to be its own PR so it can be bisected and reverted separately if it becomes a problem.

I think the change is easy, but it is much harder to determine if it is correct. I'm afraid of some code depending promise. I've glanced around and I don't _think_ anyone is using a passed-in promise other than to pass it further, but I'm still leery.

Instead of making our own singleton `VoidChannelPromise`, we should just use `channel.voidPromise()` as [it already does reuse](https://github.com/netty/netty/blob/b122faea6823d7daddfecf3e414237cf6ece28f9/transport/src/main/java/io/netty/channel/AbstractChannel.java#L409).